### PR TITLE
luminous: RGW doesn't check time skew in auth v4 http header request

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/22766
http://tracker.ceph.com/issues/23252

To Do:

- [x] wait until https://github.com/ceph/ceph/pull/20013 is merged
- [x] cherry-pick https://github.com/ceph/ceph/pull/20013 into this PR
- [x] remove DNM

### CODE REVIEW

**Comment by robbat2** [comment]
20013 is merged now, cherry-pick time?

File: src/rgw/rgw_auth_s3.cc [hunk 1]
```diff
@@ -223,8 +223,20 @@ namespace rgw {
...
 namespace s3 {
 
-/* FIXME(rzarzynski): duplicated from rgw_rest_s3.h. */
-#define RGW_AUTH_GRACE_MINS 15
+bool is_time_skew_ok(time_t t)
+{
+  auto req_tp = ceph::coarse_real_clock::from_time_t(t);
+  auto cur_tp = ceph::coarse_real_clock::now();
+
+  if (std::chrono::abs(cur_tp - req_tp) > RGW_AUTH_GRACE) {
```
Comment (cbodley) [inline src/rgw/rgw_auth_s3.cc]: oops! the make check failure is a real one. `std::chrono::abs()` is a c++17ism, so we'll need to do this instead:
```c++
  if (req_tp < cur_tp - RGW_AUTH_GRACE ||
      req_tp > cur_tp + RGW_AUTH_GRACE) {
```

Comment (smithfarm) [reply src/rgw/rgw_auth_s3.cc]: Thanks, @cbodley! Done.

**Comment by smithfarm** [comment]
Meh, my monkey-coding yielded a C++ error

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_auth_s3.cc: In function ‘int rgw::auth::s3::parse_v4_query_string(const req_info&, boost::string_view&, boost::string_view&, boost::string_view&, boost::string_view&)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_auth_s3.cc:288:24: error: ‘RGW_AUTH_GRACE_MINS’ was not declared in this scope
   if ((now_req < now - RGW_AUTH_GRACE_MINS * 60 ||
                        ^
src/rgw/CMakeFiles/rgw_a.dir/build.make:186: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/rgw_auth_s3.cc.o' failed
make[3]: *** [src/rgw/CMakeFiles/rgw_a.dir/rgw_auth_s3.cc.o] Error 1
```

**Comment by cbodley** [comment]
@smithfarm i don't think that was you - it looks like we missed a prior backport for http://tracker.ceph.com/issues/22439. i just changed that one to pending backport

**Comment by smithfarm** [comment]
@cbodley well I did have to change RGW_AUTH_GRACE to RGW_AUTH_GRACE_MINS, no?

**Comment by cbodley** [comment]
> @cbodley well I did have to change RGW_AUTH_GRACE to RGW_AUTH_GRACE_MINS, no?

nope, pr https://github.com/ceph/ceph/pull/20013 removed the `#define RGW_AUTH_GRACE_MINS 15` and added `static constexpr auto RGW_AUTH_GRACE = std::chrono::minutes{15};`

the compile error was an old reference to `RGW_AUTH_GRACE_MINS` that was removed in the pr associated with http://tracker.ceph.com/issues/22439

**Comment by smithfarm** [comment]
@cbodley OK, but you did leave off the _MINS part in your comment. The commit for 22439 cherry-picked cleanly into this PR.

**Comment by cbodley** [comment]
> @cbodley OK, but you did leave off the _MINS part in your comment.

that was intentional, because `RGW_AUTH_GRACE_MINS` no longer exists after these commits

**Comment by smithfarm** [comment]
@cbodley Could you look at the current build error, then? I'm guessing I need to replace the conditional with something like:

```
+  time_t exp = atoll(expires.data());
+  if ((exp < 1) || (exp > 7*24*60*60)) {
```

**Comment by cbodley** [comment]
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_auth_s3.cc: In function ‘bool rgw::auth::s3::is_time_skew_ok(time_t)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_auth_s3.cc:231:25: error: ‘RGW_AUTH_GRACE_MINS’ was not declared in this scope
   if (req_tp < cur_tp - RGW_AUTH_GRACE_MINS ||
                         ^
src/rgw/CMakeFiles/rgw_a.dir/build.make:186: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/rgw_auth_s3.cc.o' failed

`RGW_AUTH_GRACE_MINS` -> `RGW_AUTH_GRACE`

**Comment by smithfarm** [comment]
FACEPALM now I see, thanks @cbodley

**Comment by cbodley** [comment]
jenkins test this please

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-03-19-1611-luminous](https://trello.com/c/jFbrKQVW/277-wip-yuri3-testing-2018-03-19-1611-luminous)

**Pull request merged**